src/animation.rs

```diff
@@ -62,8 +65,21 @@ impl Animation {
             trace!("part {:?} was found", part.file);
         }
 
+        animation.validate_modes()?;
+
         Ok(animation)
     }
+
+    fn validate_modes(&self) -> Result<(), Error> {
+        // Ensure if there is any part with `Mode::Forever` it is the last one,
+        // or it will never be played.
+        let forever_part = self.parts.iter().find(|p| p.mode == Mode::Forever);
+        let last_part = self.parts.iter().last();
+        match (forever_part, last_part) {
+            (Some(forever), Some(last)) if forever != last => Err(Error::WrongModeForeverPart),
+            (..) => Ok(()),
+        }
```

Comment:
We can easily avoid the double iterator here:

```suggestion
        let mut iter = self.parts.iter();
        if let iter.find(|p| p.mode == Mode::Forever).is_sone() {
            if iter.next().is_sone() {
                return Err(Error::WrongModeForeverPart);
            }
        }

        Ok(())
```

Reply:
Nice! Indeed.

I ended doing it like:

```rust
        // Ensure if there is any part with `Mode::Forever` it is the last one,
        // or it will never be played.
        let mut iter = self.parts.iter();
        if iter.find(|p| p.mode == Mode::Forever).is_some() && iter.next().is_some() {
            return Err(Error::WrongModeForeverPart);
        }

        Ok(())
```